luda-adventure/src/app/game/types/game_object/types/positioner/positioner.rs

```diff
@@ -0,0 +1,156 @@
+        self.movement_plan =
+            MovementPlan::move_now(current_position, current_time, duration, velocity);
+        if let Some(heading) = get_heading_from_velocity(velocity) {
+            self.heading = heading;
+        }
+    }
+    pub fn get_predicted_movement_end_time(&self) -> Time {
+        self.movement_plan.get_predicted_movement_end_time()
+    }
+    pub fn predict_movement(
+        &mut self,
+        collider: &Collider,
+        collision_box_list: &[crate::app::game::CollisionBox],
+    ) {
+        let prediction_start_time = self.get_predicted_movement_end_time();
+        let prediction_start_position = self.get_position(prediction_start_time);
+        let character_collision_box = collider.get_collision_box(prediction_start_position);
+
+        let directed_movement_ended =
+            prediction_start_time == self.movement_plan.directed_movement.end_time;
+        if directed_movement_ended {
+            self.movement_plan
+                .predicted_movement_list
+                .push(super::Movement::stay_forever(
+                    prediction_start_position,
+                    prediction_start_time,
+                ));
+            return;
+        }
+
+        let first_collision = predict_collision(
+            self.movement_plan.directed_movement.velocity,
+            &character_collision_box,
+            &collision_box_list,
+            prediction_start_time,
+            self.movement_plan.directed_movement.end_time,
+            prediction_start_position,
+        );
+        match first_collision {
+            Some(first_collision) => {
+                let movement_until_first_collision = Movement {
+                    start_time: prediction_start_time,
+                    end_time: first_collision.start_time,
+                    start_position: prediction_start_position,
+                    end_position: first_collision.start_position,
+                    velocity: self.movement_plan.directed_movement.velocity,
+                };
+                self.movement_plan
+                    .predicted_movement_list
+                    .push(movement_until_first_collision);
+
+                let character_collision_box_at_first_collision =
+                    collider.get_collision_box(self.get_position(first_collision.start_time));
+                let velocity_after_first_collision = match first_collision.direction {
+                    CollisionDirection::Vertical => Xy {
+                        x: self.movement_plan.directed_movement.velocity.x,
+                        y: Per::new(0.tile(), 1.ms()),
+                    },
+                    CollisionDirection::Horizontal => Xy {
+                        x: Per::new(0.tile(), 1.ms()),
+                        y: self.movement_plan.directed_movement.velocity.y,
+                    },
+                };
+                let second_collision = predict_collision(
```

Comment:
Why this function calculates two collision at once?

Reply:
Because of float, it was ambiguous to determine whether it was in a collision.
So there was a case of getting infinitely close to the wall.
So I decided to predict post-collision movements instead of checking if they were in a collision.

Reply:
What you mean is that it would be ambiguous to determine whether it was in a collision or not because of floating number error?

Do `second_collision` mean the second collision after the first collision so it moved by one direction then collide? like a, character moves up-right and collide on top and move right and collide on right-situation?

Could it be good solution for that cases...? I'm not sure about it. I just thought saving collision status instead of calculating every time whether it is in a collision pr not - because of floating number error.

Reply:
1. Yes
2. Yes
3. Hmmmmmm. I'm not sure this will reduce the complexity yet. Let me think for one more night.

Reply:
I don't think it's going to be much. Can you explain in more detail the logic you are thinking of?

Reply:
[As I said in zulip before](https://namseent.zulipchat.com/#narrow/stream/332914-.EA.B8.B0.EC.88.A0.EC.97.B0.EA.B5.AC.EC.86.8C/topic/.28no.20topic.29/near/298384104), If I make it, I will make a queue of movement status with enum.

```rust
enum Movement {
  FreeMove {
    from: Xy,
    to: Xy
  },
  CollideAndSlideToWall {
    start: Xy,
    wall_axis: XOrY,
  },
  CollideAndStuck {
    position: Xy,
  }
}

let movement_queue: Vec<Movement> = calculate_movement(from_time, to_time);

let current_position = calculate_position(time, &movement_queue);

fn calculate_movement(from_time: Time, to_time: Time) -> Vec<Movement> {
  ...
  if let Movement(CollideAndSlideToWall { start, wall_axis }) = last_movement {
    if there_is_new_wall_so_will_be_stuck_by_them {
      queue.push(Movement::CollideAndStuck { position })
    }
  }
}
```

Reply:
What you have a trouble is that your data structures don't have enough data to calculate or `express` what you want to do.

Reply:
I think I understand the word "saving collision status".
Logic has not changed much, but the number of collision calculations has decreased.
Readability is... I don't know.